internal/app/machined/pkg/controllers/seccomp/seccomp.go

```diff
@@ -0,0 +1,169 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package seccomp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	runtimetalos "github.com/talos-systems/talos/internal/app/machined/pkg/runtime"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+	runtimeres "github.com/talos-systems/talos/pkg/machinery/resources/runtime"
+	"github.com/talos-systems/talos/pkg/machinery/resources/seccomp"
+)
+
+// ProfilesController manages v1alpha1.Stats which is the current snaphot of the machine CPU and Memory consumption.
+type ProfilesController struct {
+	V1Alpha1Mode            runtimetalos.Mode
+	SeccompProfileDirectory string
+}
+
+// Name implements controller.StatsController interface.
+func (ctrl *ProfilesController) Name() string {
+	return "seccomp.SeccompProfilesController"
+}
+
+// Inputs implements controller.StatsController interface.
+func (ctrl *ProfilesController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.StatsController interface.
+func (ctrl *ProfilesController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: seccomp.ProfileType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.StatsController interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *ProfilesController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	// initially, wait for the machine-id to be generated and /var to be mounted
+	if err := r.UpdateInputs([]controller.Input{
+		{
+			Namespace: runtimeres.NamespaceName,
+			Type:      runtimeres.MountStatusType,
+			ID:        pointer.To(constants.EphemeralPartitionLabel),
+			Kind:      controller.InputWeak,
+		},
+	}); err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		_, err := r.Get(ctx, resource.NewMetadata(runtimeres.NamespaceName, runtimeres.MountStatusType, constants.EphemeralPartitionLabel, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// in container mode EPHEMERAL is always mounted
+				if ctrl.V1Alpha1Mode != runtimetalos.ModeContainer {
+					// wait for the EPHEMERAL to be mounted
+					continue
+				}
+			} else {
+				return fmt.Errorf("error getting ephemeral mount status: %w", err)
+			}
+		}
+
+		break
+	}
+
+	// normal reconcile loop
+	if err := r.UpdateInputs([]controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
```

Comment:
currently this controller will always run on any `machineconfig` change. this is not the biggest problem, but we can avoid it by introducing an intermediate resource & controller which translate `MachineConfig` into a specific `k8s.SeccompProfileConfig` resource. This way this controller will watch `k8s.SeccompProfileConfig`, and it will not be triggered on machine config changes which do not contain changes to seccomp profiles

Reply:
fixed